docs/proposals/20230612-node-os-image-caching.md

```diff
@@ -119,29 +116,33 @@ None currently known to be needed
 #### Non-Functional Requirements
 
 ##### NFR1
-The solution should increase the performance of horizontal scaling over time as people use this feature
 
 ##### NFR2
-The solution should cause no increased risks with security or reliability
 
 ### Implementation Details/Notes/Constraints
 
-The plan is to create a Controller which implements the feature with the Node Prototype Pattern as desired.
+The plan is to modify the existing Controllers with the Node Prototype Pattern as desired. These controller additions can be added to AzureMachine Controller and AzureMachinePool Controller.
 
-These controller additions can be added to azuremachine_controller.go because that is where AzureMachines are being controlled and azuremachinepool_controller.go since that is where the AzureMachinePools are being controlled.
+An operator will be able to decide to turn the feature on or off with an environment variable on creation of a cluster, and then can update the AzureMachineTemplate and AzureMachinePool to customize how long they want the caching interval to be (see the yaml files below in this section for the caching interval).
 
-An operator will be able to decide to turn the feature on or off with a flag on creation of a cluster, and then can alter the YAML file for AzureMachineTemplate and AzureMachinePool to customize how long they want the caching interval to be.
+Example of the enviornment variable being turned on:
 
-Every time that amount of time has passed, it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a shared image gallery image, delete the snapshot, and then configure the AzureMachineTemplate specs to use that shared image gallery image.
+```
+export AZURE_OS_CACHING=true
+```
+
+The controller will maintain a timestamp, and when the current time is the chosen interval ahead or more, the controller will perform the caching. When the process is started it should go through the nodes of the cluster, choose a healthy node, shut it down, take a snapshot of it, restart it, create a shared image gallery image, delete the snapshot, and then configure the AzureMachineTemplate specs to use that shared image gallery image. After, it will store the current time as its timestamp.
```

Comment:
> The controller will maintain a timestamp

Where exactly might the timestamp be kept?

And might it be worth building in some additional assurance that the AzureMachinePool/AzureMachineTemplate object gets requeued shortly after its due time to be re-cached? The controller manager currently requeues all objects every 10 minutes by default, so it might be good enough to rely on that if the expected interval for caching is on the order of 24 hours.

Reply:
Proposed putting the timestamp on the resources themselves and 10 minutes is plenty frequent so I noted that.